pulp_2to3_migration/app/pre_migration.py

```diff
@@ -355,8 +355,46 @@ def pre_migrate_all_without_content(plan):
             # filter by repo type
             repos_to_check = plan.type_to_repo_ids[plugin_plan.type]
 
-            mongo_repo_q = mongo_Q(repo_id__in=repos_to_check)
-            mongo_repo_qs = Repository.objects(mongo_repo_q)
+            # figure out which repositories/importers/distributors were updated since the last run
+            repo_premigrated_last = Pulp2Repository.objects.aggregate(
+                Max('pulp_last_updated')
+            )['pulp_last_updated__max'] or 0
+            imp_premigrated_last = Pulp2Importer.objects.aggregate(
+                Max('pulp_last_updated')
+            )['pulp_last_updated__max'] or 0
+            dist_premigrated_last = Pulp2Distributor.objects.aggregate(
+                Max('pulp_last_updated')
+            )['pulp_last_updated__max'] or 0
+
+            is_content_added_q = mongo_Q(last_unit_added__gte=repo_premigrated_last)
+            is_content_removed_q = mongo_Q(last_unit_removed__gte=repo_premigrated_last)
+            is_new_enough_repo_q = is_content_added_q | is_content_removed_q
+            is_empty_repo_q = (
+                    mongo_Q(last_unit_added__exists=False) &
+                    mongo_Q(last_unit_removed__exists=False)
+            )
+            is_new_enough_imp_q = mongo_Q(last_updated__gte=imp_premigrated_last)
+            is_new_enough_dist_q = mongo_Q(last_updated__gte=dist_premigrated_last)
+            repo_id_q = mongo_Q(repo_id__in=repos_to_check)
+
+            updated_importers = Importer.objects(
+                repo_id_q & is_new_enough_imp_q
+            ).only('repo_id')
+            updated_imp_repos = set(imp.repo_id for imp in updated_importers)
+            updated_distributors = Distributor.objects(
+                repo_id_q & is_new_enough_dist_q
+            ).only('repo_id')
+            updated_dist_repos = set(dist.repo_id for dist in updated_distributors)
+            updated_impdist_repos = updated_imp_repos | updated_dist_repos
+            del updated_imp_repos
+            del updated_dist_repos
+
+            mongo_updated_repo_q = repo_id_q & (is_new_enough_repo_q | is_empty_repo_q)
+            mongo_updated_imp_dist_repo_q = mongo_Q(repo_id__in=updated_impdist_repos)
+
+            mongo_repo_qs = Repository.objects(
+                mongo_updated_repo_q | mongo_updated_imp_dist_repo_q
+            ).order_by('last_unit_added')
```

Comment:
no index in pulp2, maybe not a big deal since the number of repos is relatively low?

Reply:
"How many is too many" may be difficult to set a hard limit on, since the failure mode is "query takes too much memory". We'll def want to experiment to see what it takes to fail this, I would think.

Reply:
Probably fine.

Reply:
This code ran against a test-server with many (tens of thousands) of repos and didn't fail. I think we're good without worrying about a Pulp2 index.